Bump up PHP version constraint.

ADmad opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: master branch
  * Bugfix: master branch
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: master branch
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation | yes/no
| Bugfix        | yes/no
| BC Break      | yes/no
| New Feature   | yes/no
| RFC           | yes/no
| QA            | yes/no

### Description

<!--
Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE master BRANCH

- Are you adding documentation?
  - TARGET THE master BRANCH

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE master BRANCH

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE master BRANCH

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE develop BRANCH

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE develop BRANCH
-->

ADmad committed:
Bump up PHP version constraint.

Signed-off-by: GitHub <noreply@github.com>

ADmad committed:
Update testsuite matrix on TravisCI

Signed-off-by: GitHub <noreply@github.com>

ADmad committed:
Fix phpunit version constraint.

Signed-off-by: GitHub <noreply@github.com>

ADmad commented:
Travis is AWOL.

weierophinney commented:
> Travis is AWOL.

We're aware; we've used up our allotment of credit hours for the month. I've got a support request in; in the meantime, we'll be attempting to verify manually before merge.

ADmad commented:
You'll have to switch to Github Actions (or another CI provider) eventually :slightly_smiling_face:.

weierophinney commented:
> You'll have to switch to Github Actions (or another CI provider) eventually .

Already planned: https://github.com/laminas/technical-steering-committee/blob/main/meetings/minutes/2020-11-02-TSC-Minutes.md#github-actions

ADmad commented:
If you want I can add a simple new workflow to run the testsuite on Github Action. You can later remove it or improve upon.

weierophinney commented:
> If you want I can add a simple new workflow to run the testsuite on Github Action. You can later remove it or improve upon.

Would rather wait, as it will be easier to script adding the workflow(s) to repositories later if they don't already have ad-hoc changes.

ADmad commented:
> Would rather wait, as it will be easier to script adding the workflow(s) to repositories later if they don't already have ad-hoc changes.

Sorry, I added one. After seeing the test suite passing on PHP 8, I removed it again :slightly_smiling_face:.

weierophinney commented on composer.json:
```diff
@@ -27,7 +27,7 @@
     },
     "require": {
-        "php": "^7.1",
+        "php": "^7.2 || ~8.0.0",
```
We can, and should, bump this to 7.3, as 7.2 goes out of support in 11 days.

boesing replied:
Actually, its PHP 7.3 which goes out of support in 11 days.
https://www.php.net/supported-versions

ADmad replied:
> Actually, its PHP 7.3 which goes out of support in 11 days.

True, but there's no need to drop support for PHP 7.3 just yet.

boesing replied:
> True, but there's no need to drop support for PHP 7.3 just yet.

Yup, just wanted to fix that message as it states that PHP 7.3 might be supported for another year which aint true.

weierophinney replied:
@boesing Sorry, let me clarify: PHP 7.2 goes end-of-life in 11 days. PHP 7.3 continues to receive security patches for another year. Our security window is for versions that receive updates of any kind, so it's those getting both active support and those getting security patches. This is why I suggested bumping the minimum to 7.3.

weierophinney commented on composer.json:
```diff
@@ -36,7 +36,7 @@
         "laminas/laminas-coding-standard": "~1.0.0",
         "laminas/laminas-diactoros": "^1.7 || ^2.1.1",
-        "phpunit/phpunit": "^7.0.2"
+        "phpunit/phpunit": "^8.5 || ^9.1"
```
If we bump to PHP 7.3, we can bump this to just `^9.3`.

weierophinney commented on composer.json:
```diff
@@ -36,7 +36,7 @@
     "require-dev": {
         "laminas/laminas-coding-standard": "~1.0.0",
         "laminas/laminas-diactoros": "^1.7 || ^2.1.1",
```
We can remove the 1.7 constraint here.

weierophinney reviewed (changes requested):
I've flagged a few changes; I'll go ahead and make them and push them back after testing on each of the target versions locally.

weierophinney committed:
qa: update dependency constraints

- Bump minimum required PHP to 7.3.
- Bump minimum supported PHPUnit to 9.3.
- Bump minimum supported Diactoros to 2.1.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

ADmad commented:
> I'll go ahead and make them and push them back after testing on each of the target versions locally.

Cool, I'll await your updates.

weierophinney committed:
qa: update test suite to run under PHPUnit 9.3+

- Updates phpunit schema
- Removes usage of Prophecy, replacing with native PHPUnit mocks.
- Introduces MockStreamHelper, to make handling stream mocks for testing
  the stream emitter easier.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
Adds CHANGELOG entry for #7

Signed-Off-By: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney commented:
I took the liberty of removing Prophecy mocks for native PHPUnit mocks... which required a fair bit of change to the `SapiStreamEmitterTest`, as it had a rather convoluted setup. I extracted much of it to a `MockStreamHelper` that provides state and callbacks that a mock stream instance can use, which leads to more explicit setup. All the original test cases continue to pass on both lowest and latest dependencies locally, so I'm going to go ahead and merge.
